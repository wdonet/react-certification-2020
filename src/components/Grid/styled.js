import styled from 'styled-components';

const List = styled.ul`
  display: grid;
  grid-template-columns: repeat(auto-fill, minmax(320px, 1fr));
  grid-gap: 1rem;
`;

const GridContainer = styled.div`
  padding: 2rem;
`;

export { GridContainer, List };
