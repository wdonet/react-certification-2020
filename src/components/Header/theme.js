const base = {
  easeOutBack: 'cubic-bezier(0.34, 1.56, 0.64, 1)',
  colorWhite: 'rgb(255, 255, 255)',
  colorBlack: 'rgb(0, 0, 0)',
};

const dark = {
  id: 'dark',
  ...base,
  backgroundColor: '#333',
  textColor: 'black',
};

const light = {
  id: 'light',
  ...base,
  backgroundColor: '#666',
  textColor: 'white',
};

export const theme = { dark, light };
