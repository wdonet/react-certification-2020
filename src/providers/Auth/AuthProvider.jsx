import React, { useState, useEffect, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { loginUser } from '../../mocks/loginApi';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

function AuthProvider({ authenticatedJest, children }) {
  const [authenticated, setAuthenticated] = useState(authenticatedJest);

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setAuthenticated(isAuthenticated || authenticatedJest);
  }, []);

  const login = async ({ username, password }) => {
    try {
      const authorized = await loginUser({ username, password });
      if (authorized) {
        setAuthenticated(true);
        storage.set(AUTH_STORAGE_KEY, true);
      }
      return { success: true };
    } catch (error) {
      return { error };
    }
  };

  const logout = useCallback(() => {
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, false);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  );
}

export { AuthContext };
export default AuthProvider;
