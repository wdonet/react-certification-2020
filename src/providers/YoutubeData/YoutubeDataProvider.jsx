import React, { useEffect, useReducer } from 'react';
import { getVideosByQuery, getVideoDetails, getRelatedVideos } from './api';
import reducer from './reducer';
import { setSearchTerm, setVideos } from './actions';

const YoutubeDataContext = React.createContext(null);

const YoutubeDataProvider = ({ iframeAPIReady, children }) => {
  const [state, dispatch] = useReducer(reducer, {
    videos: [],
    searchTerm: '',
  });

  const { videos, searchTerm } = state;

  const fetchVideos = async (query) => {
    try {
      const newVideos = await getVideosByQuery(query);
      setVideos(dispatch, newVideos);
    } catch (error) {
      console.error(error);
    }
  };

  const fetchVideoDetails = async (videoId) => {
    try {
      const videoDetails = await getVideoDetails(videoId);
      return videoDetails;
    } catch (error) {
      console.error(error);
    }
  };

  const fetchRelatedVideos = async (videoId) => {
    try {
      const relatedVideos = await getRelatedVideos(videoId);
      return relatedVideos;
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    fetchVideos();
  }, []);

  return (
    <YoutubeDataContext.Provider
      value={{
        videos,
        fetchVideos,
        iframeAPIReady,
        fetchVideoDetails,
        fetchRelatedVideos,
        searchTerm,
        setSearchTerm: (search) => setSearchTerm(dispatch, search),
      }}
    >
      {children}
    </YoutubeDataContext.Provider>
  );
};

YoutubeDataProvider.propTypes = {};

export { YoutubeDataContext };
export default YoutubeDataProvider;
