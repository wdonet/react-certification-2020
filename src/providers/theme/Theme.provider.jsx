// React
import React, { useState, useEffect, useContext, useCallback } from 'react';
// Utils
import { THEME_SELECTED } from '../../utils/constants';
import { storage } from '../../utils/storage';
// Themes
import light from '../../theme/light';
import dark from '../../theme/dark';

const ThemeContext = React.createContext({ theme: light, selected: false });

const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error(`Can't use "useTheme" without an ThemeProvider!`);
  }
  return context;
};

const ThemeBaseProvider = ({ children }) => {
  const [theme, setTheme] = useState(light);
  const [selected, setSelected] = useState(false);

  useEffect(() => {
    const lastTheme = storage.get(THEME_SELECTED);
    const whichTheme = lastTheme ? dark : light;
    setSelected(lastTheme);
    setTheme(whichTheme);
  }, []);

  const toggleTheme = useCallback((toggle) => {
    setTheme(toggle ? dark : light);
    storage.set(THEME_SELECTED, toggle);
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, selected, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export { useTheme, ThemeContext };
export default ThemeBaseProvider;
