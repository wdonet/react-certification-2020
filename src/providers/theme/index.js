export { default, useTheme } from './Theme.provider';
