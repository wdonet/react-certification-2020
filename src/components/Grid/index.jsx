import React from 'react';
import { GridContainer, List } from './styled';

const Grid = ({ gridItems }) => {
  return (
    <GridContainer>
      <List> {gridItems}</List>
    </GridContainer>
  );
};

export default Grid;
