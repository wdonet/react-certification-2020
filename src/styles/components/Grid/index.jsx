import React from 'react'
import { StyledGrid } from './styled'

const Grid = props => <StyledGrid data-testid="Grid">{props.children}</StyledGrid>

Grid.propTypes = {}

export default Grid
