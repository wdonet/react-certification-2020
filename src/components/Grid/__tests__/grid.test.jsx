import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';

import Grid from '../index';
import AuthProvider from '../../../providers/Auth';
import YoutubeDataProvider from '../../../providers/YoutubeData';

import { lightTheme } from '../../../theme';

describe('Grid', () => {
  test('renders correctly', () => {
    const items = [<div key="1">Hola</div>, <div key="2">Adios</div>];
    const tree = renderer
      .create(
        <BrowserRouter>
          <AuthProvider>
            <YoutubeDataProvider>
              <ThemeProvider theme={lightTheme}>
                <Grid gridItems={items} />
              </ThemeProvider>
            </YoutubeDataProvider>
          </AuthProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
