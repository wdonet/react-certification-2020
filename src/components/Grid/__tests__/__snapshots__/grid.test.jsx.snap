// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders correctly 1`] = `
.c1 {
  display: grid;
  grid-template-columns: repeat(auto-fill,minmax(320px,1fr));
  grid-gap: 1rem;
}

.c0 {
  padding: 2rem;
}

<div
  className="c0"
>
  <ul
    className="c1"
  >
     
    <div>
      Hola
    </div>
    <div>
      Adios
    </div>
  </ul>
</div>
`;
